java/src/test/java/actorworkflow/CustomerLoyaltyTest.java

```diff
@@ -0,0 +1,176 @@
+        WorkflowOptions workflowOptions =
+                WorkflowOptions.newBuilder()
+                        .setTaskQueue(testWorkflowRule.getTaskQueue())
+                        .setWorkflowId(Shared.WORKFLOW_ID_FORMAT.formatted("host"))
+                        .setWorkflowIdReusePolicy(
+                                WorkflowIdReusePolicy.WORKFLOW_ID_REUSE_POLICY_REJECT_DUPLICATE)
+                        .build();
+        CustomerLoyaltyWorkflow workflow =
+                testWorkflowRule
+                        .getWorkflowClient()
+                        .newWorkflowStub(CustomerLoyaltyWorkflow.class, workflowOptions);
+
+        var customer = new Customer("host", "", 0, StatusTier.STATUS_TIERS.get(4), new ArrayList<>());
+        WorkflowClient.start(workflow::customerLoyalty, customer);
+
+        int order = 0;
+        var guest = new Customer("guest");
+        testWorkflowRule.getTestEnvironment()
+                .registerDelayedCallback(Duration.ofSeconds(order++),
+                        () -> workflow.inviteGuest(guest));
+
+        testWorkflowRule.getTestEnvironment()
+                .registerDelayedCallback(Duration.ofSeconds(order++),
+                        () -> workflow.inviteGuest(guest));
+
+        testWorkflowRule.getTestEnvironment()
+                .registerDelayedCallback(Duration.ofSeconds(order++),
+                        workflow::cancelAccount);
+
+        CustomerLoyaltyWorkflow child = testWorkflowRule
+                .getWorkflowClient()
+                .newWorkflowStub(CustomerLoyaltyWorkflow.class,
+                        WorkflowOptions.newBuilder()
+                                .setTaskQueue(testWorkflowRule.getTaskQueue())
+                                .setWorkflowId(Shared.WORKFLOW_ID_FORMAT.formatted(guest.customerId()))
+                                .build());
+
+        testWorkflowRule.getTestEnvironment().registerDelayedCallback(Duration.ofSeconds(order++), () -> {
+            // "start" the workflow, to make sure we have the current execution, but expect it to throw
+            try {
+                WorkflowClient.start(child::customerLoyalty, guest);
+            } catch (WorkflowExecutionAlreadyStarted ignored) {
+            }
+            child.cancelAccount();
+        });
+
+        testWorkflowRule.getTestEnvironment().sleep(Duration.ofSeconds(order + 1));
+
+        verify(activities, times(1))
+                .sendEmail(EmailStrings.EMAIL_GUEST_INVITED);
+        verify(activities, times(1))
+                .sendEmail(EmailStrings.EMAIL_GUEST_MIN_STATUS.formatted(StatusTier.STATUS_TIERS.get(3).name()));
+
+        testWorkflowRule.getTestEnvironment().shutdown();
+    }
+
+    @Test(expected = Test.None.class)
+    public void testSimpleReplay() throws Exception {
+        ClassLoader classLoader = getClass().getClassLoader();
+        File file = new File(Objects.requireNonNull(classLoader.getResource("simple_replay.json")).getFile());
```

Comment:
instead of having a hard-coded history json you could run your wf exec in test and after it completes get 
its event history and then run it through replayer

Reply:
I'm not sure I follow -- wouldn't that test never fail if I run the wf and then immediately run that history through the replayer? My understanding was that using a static history download would mean that the replay was actually testing against a genuine past exec.

Reply:
see this test for reference if that helps
https://github.com/temporalio/samples-java/blob/main/core/src/test/java/io/temporal/samples/hello/HelloActivityReplayTest.java#L100